### PR TITLE
Google Drive Mounting with elFinder

### PR DESCRIPTION
**Pull request by rajasharma27**
Sir, 
this class working perfectly.
please add in your elFinder

### CODE REVIEW

File: php/elFinderVolumeGoogleDrive.class.php [hunk 1]
```diff
@@ -0,0 +1,1659 @@
...
+	/**
+	 * Constructor
+	 * Extend options with required fields
+	 *
+	 * @return void
+	 * @author Dmitry (dio) Levashov
+	 * @author Cem (DiscoFever)
+	 **/
+	public function __construct() {	
+		@ include_once 'google-api-php-client/vendor/autoload.php';
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Please remove this line. You must be add an option for google-api-php-client auto loader.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 2]
```diff
@@ -0,0 +1,1659 @@
...
+			'client_id'       	=> '',
+			'client_secret'    	=> '',				
+			'accessToken'       => '',
+			'accessTokenSecret' => '',
+			'googledriveUid'    => '',
+			'root'				=> 'GoogleDrive.com',						
+			'path'              => '/',
+			'childpath'         => '',
+			'separator'         => '/',			
+			'treeDeep'          => 0,
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Please remove this line. You must be support `treeDeep` option.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 3]
```diff
@@ -0,0 +1,1659 @@
...
+			'tmbPath'           => '../files/.tmb',
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Please remove next 3 lines. ('tmbPath', 'tmbURL', 'tmpPath')

File: php/elFinderVolumeGoogleDrive.class.php [hunk 4]
```diff
@@ -0,0 +1,1659 @@
...
+			'tmbURL'            => dirname($_SERVER['PHP_SELF']) . '/../files/.tmb',
+			'tmpPath'           => '../files/.tmp',
+			'getTmbSize'        => 'large', // small: 32x32, medium or s: 64x64, large or m: 128x128, l: 640x480, xl: 1024x768
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Please remove next 3 lines and you must be clean up code of related this lines.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 5]
```diff
@@ -0,0 +1,1659 @@
...
+		}
+		if (!$this->metaCache && $this->tmp) {
+			$this->metaCache = $this->tmp;
+		}
+		
+		if (!$this->metaCache) {
+			return $this->setError('Cache dirctory (metaCachePath or tmp) is require.');
+		}
+		
+		if (is_null($this->options['syncChkAsTs'])) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: GoogleDrive driver is not support $this->options['syncChkAsTs'] = true. You must be set $this->options['syncChkAsTs'] = false.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 6]
```diff
@@ -0,0 +1,1659 @@
...
+						$client->setRedirectUri($this->getConnectorUrl().'?cmd=netmount&protocol=googledrive&host=1');						
+						//$client->setScopes([Google_Service_Oauth2::USERINFO_PROFILE , Google_Service_Drive::DRIVE]);
+						$client->setScopes([Google_Service_Drive::DRIVE]);
+						if (! empty($options['offline'])) {
+                        	$client->setApprovalPrompt('force');
+                        	$client->setAccessType('offline');
+                    	}						
+					
+						$service = new Google_Service_Drive($client);
+    					$this->oauth = $client->createAuthUrl();
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: assignment to $this->oauth is unnecessary

File: php/elFinderVolumeGoogleDrive.class.php [hunk 7]
```diff
@@ -0,0 +1,1659 @@
...
+	 * @return boolean
+	 */
+	public function netunmount($netVolumes, $key) {	
+		$count = 0;
+		$googledriveUid = '';
+		if (isset($netVolumes[$key])) {
+			$googledriveUid = $netVolumes[$key]['googledriveUid'];
+		}
+		foreach($netVolumes as $volume) {
+			if (@$volume['host'] === 'googledrive' && @$volume['googledriveUid'] === $googledriveUid) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 8]
```diff
@@ -0,0 +1,1659 @@
...
+			
+		}
+
+		$this->rootName = $this->options['alias'];
+				
+		$this->googledriveUid = $this->options['googledriveUid'];
+		$this->tmbPrefix = 'googledrive'.base_convert($this->googledriveUid, 10, 32);
+		
+		if (!empty($this->options['tmpPath'])) {
+			if ((is_dir($this->options['tmpPath']) || @mkdir($this->options['tmpPath'])) && is_writable($this->options['tmpPath'])) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 9]
```diff
@@ -0,0 +1,1659 @@
...
+		}
+		if (!$this->tmp && is_writable($this->options['tmbPath'])) {
+			$this->tmp = $this->options['tmbPath'];
+		}
+		if (!$this->tmp && ($tmp = elFinder::getStaticVar('commonTempPath'))) {
+			$this->tmp = $tmp;
+		}
+		
+		if (!empty($this->options['metaCachePath'])) {
+			if ((is_dir($this->options['metaCachePath']) || @mkdir($this->options['metaCachePath'])) && is_writable($this->options['metaCachePath'])) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 10]
```diff
@@ -0,0 +1,1659 @@
...
+
+	/**
+	 * Close opened file
+	 *
+	 * @param  resource  $fp  file pointer
+	 * @return bool
+	 * @author Dmitry (dio) Levashov
+	 **/
+	protected function _fclose($fp, $path='') {	
+		@fclose($fp);
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 11]
```diff
@@ -0,0 +1,1659 @@
...
+		if ($path) {
+			@unlink($this->getTempFile($path));
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 12]
```diff
@@ -0,0 +1,1659 @@
...
+	 * @param  string  $content  new file content
+	 * @return bool
+	 * @author Dmitry (dio) Levashov
+	 **/
+	protected function _filePutContents($path, $content) {	
+		$res = false;
+		
+		if ($local = $this->getTempFile($path)) {
+			if (@file_put_contents($local, $content, LOCK_EX) !== false
+			&& ($fp = @fopen($local, 'rb'))) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 13]
```diff
@@ -0,0 +1,1659 @@
...
+				clearstatcache();
+				$res = $this->_save($fp, $path, '', array());
+				@fclose($fp);
+			}
+			file_exists($local) && @unlink($local);
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not use the `@` to hide the error.

**Comment by nao-pon** [comment]
And I recommend replace from `$this->oauth` to  `$this->client`.

**Comment by rajasharma27** [comment]
ok sir,
I am removing all these things

File: php/elFinderVolumeGoogleDrive.class.php [hunk 14]
```diff
@@ -0,0 +1,1659 @@
...
+				unlink($tmb);
+			}
+		}
+		if((int)$this->session->get('DriveCounter')>1){
+			$this->session->set('DriveCounter', (int)$this->session->get('DriveCounter')-1);
+		}else{	
+			$client = new Google_Client();
+																		   			
+			$client->setAccessToken($this->session->get('elFinderGoogleDriveAuthTokens'));									
+			$client->revokeToken();
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: revokeToken() is broken anothor fixed GoogleDrive volume drive with refreshToken. I think `$client->revokeToken()` is unnecessary.

**Comment by rajasharma27** [comment]
sir,
when I remove these line showing error in mounting
'tmbPath'           => '../files/.tmb',
'tmbURL'            => dirname($_SERVER['PHP_SELF']) . '/../files/.tmb',
'tmpPath'           => '../files/.tmp',

**Comment by rajasharma27** [comment]
error

**Comment by nao-pon** [comment]
@rajasharma27 

(Is a machine translation by Google Translate.)

I have to check the code, but it is ambiguous part in many cases. For example, it does not seem to function properly as a fixed drive. In addition, there is such as unnecessary variable definition and options in the code.

To add to a project, you need to make it work without any problems in a variety of conditions.

Currently, because GoogleDrive driver using FlySystem there is already, I priority to maintain the code for this class is low. Thus, there is a need that you raise the more complete of this code.

**Comment by nao-pon** [comment]
- [elFinder/elFinderVolumeGoogleDrive.class.php at gdrive · nao-pon/elFinder](https://github.com/nao-pon/elFinder/blob/gdrive/php/elFinderVolumeGoogleDrive.class.php)
For your reference. Still, it is in the process of being...

**Comment by rajasharma27** [comment]
ok sir,
after your change, missing google api client when "netmountPrepare"

if (! $this->service) {		
            if ($this->options['googleApiClient'] && ! class_exists('Google_Client')) {			
                include_once($this->options['googleApiClient']);				
            }
			if (! class_exists('Google_Client', false)) {			
                return $this->setError('Class Google_Client not found.');
            }
		}

**Comment by rajasharma27** [comment]
this function not work properly in netmountPrepare
if (empty($this->options['googleApiClient']) && defined('ELFINDER_GOOGLEDRIVE_GOOGLEAPICLIENT')) {
            $this->options['googleApiClient'] = ELFINDER_GOOGLEDRIVE_GOOGLEAPICLIENT;
			include_once($this->options['googleApiClient']);		
        }

**Comment by rajasharma27** [comment]
using this function in 
__construct()

**Comment by rajasharma27** [comment]
Sir,
now working perfectly

**Comment by rajasharma27** [comment]
Sir,
please add in your elFinder

**Comment by rajasharma27** [comment]
Sir,
Update and Reduce code
class working perfectly

**Comment by nao-pon** [comment]
@rajasharma27 Content URL that elFinder returns should have been published.

**Comment by rajasharma27** [comment]
Sir,
I am not getting

**Comment by rajasharma27** [comment]
Content URL that elFinder returns should have been published.
that mean
$stat['url']
$stat['width']
$stat['height']
disable form parseRaw  and use function _dimensions() and getContentUrl()

Sir,
please help

**Comment by nao-pon** [comment]
@rajasharma27 Please see nao-pon/flysystem-google-drive [`getUrl`](https://github.com/nao-pon/flysystem-google-drive/blob/master/src/GoogleDriveAdapter.php#L566-L586) and [`publish`](https://github.com/nao-pon/flysystem-google-drive/blob/master/src/GoogleDriveAdapter.php#L663-L688).

**Comment by rajasharma27** [comment]
thank
updating class for publish contents

**Comment by rajasharma27** [comment]
Sir,
update permissions for specified path or Content

**Comment by nao-pon** [comment]
@rajasharma27 Only if you request the content URL should be in public. And I recommend to publish permission `withLink` to `true` or configurable permission mode.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 15]
```diff
@@ -0,0 +1,1635 @@
...
+
+        $stat['rev']        = isset($raw['id']) ? $raw['id'] : 'root';
+        $stat['name']        = isset($raw['name']) ? $raw['name'] : '';
+        $stat['mime']        = $raw['mimeType'] == self::DIRMIME ? 'directory' :
+                              ($raw['mimeType'] == 'image/bmp' ? 'image/x-ms-bmp' :  $raw['mimeType']);
+        $stat['size']        = $raw['mimeType'] == self::DIRMIME ? 0 : (int)$raw['size'];
+        $stat['ts']            = isset($raw['modifiedTime']) ? strtotime($raw['modifiedTime']) : $_SERVER['REQUEST_TIME'];
+        $stat['dirs']        = $raw['mimeType'] == self::DIRMIME ? 1 : 0;
+        
+        if (!empty($raw['webContentLink']) && (explode('/', $raw['mimeType'])[0] =='image' || explode('/', $raw['mimeType'])[0] =='video')) {
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: If item is published return content url else return '1'.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 16]
```diff
@@ -0,0 +1,1635 @@
...
+    
+    /**
+    * Return content URL
+    *
+    * @param string  $hash  file hash
+    * @param array $options options
+    * @return array
+    * @author Naoki Sawada
+    **/
+    public function getContentUrl($hash, $options = array())
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do publish before return content url.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 17]
```diff
@@ -0,0 +1,1635 @@
...
+                      'data' => strlen($data) == 0 ? ' ' : $data,
+                      'mimeType' => $mimeType,
+                      'uploadType' => 'media'
+                ));
+            }
+        } catch (Exception $e) {
+            return $this->setError('GoogleDrive error: '.$e->getMessage());
+        }
+        $path = $this->_normpath(dirname($path).'/'.$file->getId());
+        $this->publish($path);
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not publish.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 18]
```diff
@@ -0,0 +1,1635 @@
...
+            $files->setName($name);
+                        
+            //Set the Parent id			
+            $targetDir == '/' ? $parentId = 'root' : $parentId = basename($targetDir);
+            $files->setParents(array($parentId));
+            
+            $file = $this->service->files->copy(basename($source), $files, ['fields' => self::FETCHFIELDS_GET]);
+            $itemId = $file->id;
+            
+            if($this->publish($itemId) == false){
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Do not publish.

**Comment by rajasharma27** [comment]
Sir,
please check and suggest

**Comment by rajasharma27** [comment]
Sir,
when I use resize and rotate image, it create new thumbnail in tmb path but not refresh thumbnail.
showing console error in chrome
jquery.min.js:3 Resource interpreted as Document but transferred with MIME type image/jpeg:

**Comment by rajasharma27** [comment]
Sir,
cover all points 
but "Do publish before return content url." comment not close.
please help

**Comment by rajasharma27** [comment]
Sir,
please add this class

**Comment by nao-pon** [comment]
There are API calls that were duplicated by one HTTP request. This will cause a decrease in performance.

for example...
*open folder*
```
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(123) "files(id,name,mimeType,modifiedTime,parents,permissions,size,imageMediaMetadata(height,width),webContentLink),nextPageToken"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
```

----

*mkfile*
```
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(14) "files(id,name)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
```

----

*cwd reload*
```
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(23) "files(id,name,mimeType)"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
array(4) {
  ["fields"]=>
  string(123) "files(id,name,mimeType,modifiedTime,parents,permissions,size,imageMediaMetadata(height,width),webContentLink),nextPageToken"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(59) "trashed=false and "0ByGZEnvnTbPbY3pra3RUbkdPbDg" in parents"
}
array(4) {
  ["fields"]=>
  string(123) "files(id,name,mimeType,modifiedTime,parents,permissions,size,imageMediaMetadata(height,width),webContentLink),nextPageToken"
  ["pageSize"]=>
  int(1000)
  ["spaces"]=>
  string(5) "drive"
  ["q"]=>
  string(35) "trashed=false and "root" in parents"
}
```

File: php/elFinderVolumeGoogleDrive.class.php [hunk 19]
```diff
@@ -0,0 +1,1606 @@
...
+    /**
+    * Recursive files search
+    *
+    * @param  string  $path   dir path
+    * @param  string  $q      search string
+    * @param  array   $mimes
+    * @return array
+    * @author Naoki Sawada
+    **/
+    protected function doSearch($path, $q, $mimes)
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: `doSearch` should be searching for descendants directories recursively.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 20]
```diff
@@ -0,0 +1,1606 @@
...
+        $stat = array();
+
+        $stat['rev']        = isset($raw['id']) ? $raw['id'] : 'root';
+        $stat['name']        = isset($raw['name']) ? $raw['name'] : '';
+        $stat['mime']        = $raw['mimeType'] == self::DIRMIME ? 'directory' :
+                              ($raw['mimeType'] == 'image/bmp' ? 'image/x-ms-bmp' :  $raw['mimeType']);
+        $stat['size']        = $raw['mimeType'] == self::DIRMIME ? 0 : (int)$raw['size'];
+        $stat['ts']            = isset($raw['modifiedTime']) ? strtotime($raw['modifiedTime']) : $_SERVER['REQUEST_TIME'];
+        $stat['dirs']        = $raw['mimeType'] == self::DIRMIME ? 1 : 0;
+        $stat['url'] = '1';
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: If the item is published you should set the content URL.

File: php/elFinderVolumeGoogleDrive.class.php [hunk 21]
```diff
@@ -0,0 +1,1606 @@
...
+     * @param string $path with id
+     * @return array googledrive metadata
+     */
+     
+    private function chkDBdat($path)
+    {
+        //$files = new Google_Service_Drive_DriveFile();
+
+        if ($path == $this->root) {
+            return $this->root;
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: The return value is not an array.

**Comment by rajasharma27** [comment]
sir, 
I am facing problem when recursive folder search

**Comment by rajasharma27** [comment]
sir,
doSearch update with recursively,

**Comment by rajasharma27** [comment]
Sir,
I am working on fast open folder and mkfile.
what is the main location from pick the $path
please help.

**Comment by rajasharma27** [comment]
Sir,
reduce time in open folder, mkfile, cwd reload
please consider

**Comment by rajasharma27** [comment]
Sir,
 reduce time in open folder, mkfile, cwd reload
 please consider

**Comment by rajasharma27** [comment]
Sir,
 when I trying to upload files  _savefile () function return every file 4-5 times
where is the problem
please help

File: php/elFinderVolumeGoogleDrive.class.php [hunk 22]
```diff
@@ -0,0 +1,1549 @@
...
+    {
+        $stat = array();
+
+        $stat['rev']        = isset($raw['id']) ? $raw['id'] : 'root';
+        $stat['name']        = isset($raw['name']) ? $raw['name'] : '';
+        $stat['mime']        = $raw['mimeType'] == self::DIRMIME ? 'directory' :
+                              ($raw['mimeType'] == 'image/bmp' ? 'image/x-ms-bmp' :  $raw['mimeType']);
+        $stat['size']        = $raw['mimeType'] == self::DIRMIME ? 0 : (int)$raw['size'];
+        $stat['ts']            = isset($raw['modifiedTime']) ? strtotime($raw['modifiedTime']) : $_SERVER['REQUEST_TIME'];
+        $stat['dirs']        = $raw['mimeType'] == self::DIRMIME ? 1 : 0;
```
Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: This is not good idea. Please remove this line, And to make detect `dirs` correctly.

**Comment by nao-pon** [comment]
When I edit the text file and save it, ItemID name of the file will be newly created.

File: php/elFinderVolumeGoogleDrive.class.php (hunk 22 above)
Comment (rajasharma27) [inline php/elFinderVolumeGoogleDrive.class.php]: sir, suggest me.
$stat['dirs'] =0
        if($raw['mimeType'] == self::DIRMIME){
			$stat['dirs'] = 1;
		}

Comment (nao-pon) [inline php/elFinderVolumeGoogleDrive.class.php]: Value of `$stat['dirs']` should be set to `0` if the target directory has not any child directories. Or if it is unknown, it must not be set to the value and it must be set in `_subdirs()` correctly.

Comment (rajasharma27) [inline php/elFinderVolumeGoogleDrive.class.php]: ok
I am updating _subdirs() function for checking subfolders

File: php/elFinderVolumeGoogleDrive.class.php (hunk 19 above)
Comment (rajasharma27) [inline php/elFinderVolumeGoogleDrive.class.php]: search recursively done

**Comment by rajasharma27** [comment]
sir, please add with elfinder

**Comment by nao-pon** [comment]
@rajasharma27 I do not have enough time to perform the verification over the future of the volume driver of your requests. Can you support to continue over the future?

**Comment by rajasharma27** [comment]
yes i'll support in future

**Comment by nao-pon** [comment]
@rajasharma27 I got some notice errors.

```
NOTICE: Undefined index: url in elFinderVolumeGoogleDrive.class.php line 176.
NOTICE: Undefined offset: 0 in elFinderVolumeGoogleDrive.class.php line 1451.
NOTICE: Undefined index: mime in elFinderVolumeGoogleDrive.class.php line 1490.
```

Please make sure there is no potential problem with one of these errors. And Please correct so that the error is not reported.

**Comment by nao-pon** [comment]
@rajasharma27 I recommend to capture PHP errors like this way.

```php
function debugLog() {
	if (elFinder::$phpErrors) {
		$o = join("\n", elFinder::$phpErrors) . "\n";
		file_put_contents('debuglog.txt', $o, FILE_APPEND);
	}
}

$connectorOpts = array(
	'bind' => array(
		'*' => array('debugLog')
	),
	// ... and more options
);
```

**Comment by rajasharma27** [comment]
sir,
not showing these error on my system.

**Comment by nao-pon** [comment]
@rajasharma27 Did you set PHP error_reporting to `error_reporting(-1);`?

**Comment by rajasharma27** [comment]
ok, now showing php error

**Comment by rajasharma27** [comment]
Sir,
Successfully remove PHP error

**Comment by nao-pon** [comment]
@rajasharma27 There is a problem. It can not be re-authentication when the access token time expires.

gdrive_auth_problem

**Comment by nao-pon** [comment]
@rajasharma27 Now, I'm touching this code. I'll fix it.

**Pull request merged**